### PR TITLE
Allow UDP for kubectl create service.

### LINKED ISSUE #47337: "kubectl create service" should allow UDP port
"kubectl create service" only supports TCP port. We should allow it to support UDP port.
`--tcp=[]: Port pairs can be specified as '<port>:<targetPort>'.`

Two possible ways to modify flag. I'd prefer the latter one.
- add a flag `--udp` which is similar to `--tcp`
- change flag `--tcp` to `--ports` which has format `<protocol>:<port>:<targetPort>`

Do I miss anything?

/cc @thockin @adohe 
/sig network
/sig cli
/assign @xiangpengzhao

### PR DESCRIPTION
**Pull request by sandoracs**
**What this PR does / why we need it**:
TCP is hardcoded in 'kubectl create service' however UDP should be supported too.
**Which issue this PR fixes**: 
fixes #47337

**Special notes for your reviewer**:

**Release note**:

```release-note
```

### CODE REVIEW

**Comment by CaoShuFeng** [comment]
I personally prefer adding a new --udp. Because this break backword compatibility.

**Comment by sandoracs** [comment]
I see. I can rewrite to use '--udp' if it is the preferred way.

**Comment by xiangpengzhao** [comment]
Glad to see this PR! Thanks @sandoracs !
TBH, I prefer the `--port` implementation in this PR and we'd better also mark `--tcp` as deprecated in order not to break compatibility.

@thockin WDYT?

File: pkg/kubectl/cmd/create_service_test.go [hunk 1]
```diff
@@ -48,7 +48,7 @@ func TestCreateService(t *testing.T) {
 	buf := bytes.NewBuffer([]byte{})
 	cmd := NewCmdCreateServiceClusterIP(f, buf)
 	cmd.Flags().Set("output", "name")
-	cmd.Flags().Set("tcp", "8080:8000")
+	cmd.Flags().Set("ports", "tcp:8080:8000")
```
Comment (xiangpengzhao) [inline pkg/kubectl/cmd/create_service_test.go]: refactor the test case to table-driven tests for testing multiple scenarios (tcp/udp/invalid) ?

File: pkg/kubectl/cmd/create_service.go [hunk 2]
```diff
@@ -52,20 +52,20 @@ var (
...
     # Create a new clusterIP service named my-cs
-    kubectl create service clusterip my-cs --tcp=5678:8080
+    kubectl create service clusterip my-cs --ports=tcp:5678:8080
 
     # Create a new clusterIP service named my-cs (in headless mode)
     kubectl create service clusterip my-cs --clusterip="None"`))
 )
 
 func addPortFlags(cmd *cobra.Command) {
-	cmd.Flags().StringSlice("tcp", []string{}, "Port pairs can be specified as '<port>:<targetPort>'.")
```
Comment (xiangpengzhao) [inline pkg/kubectl/cmd/create_service.go]: mark flag `tcp` as deprecated instead of removing it.

File: pkg/kubectl/service_basic.go [hunk 3]
```diff
@@ -209,18 +210,20 @@ func (s ServiceCommonGeneratorV1) StructuredGenerate() (runtime.Object, error) {
...
+	for _, portsString := range s.Ports {
+		port, targetPort, err := parsePorts(portsString)
 		if err != nil {
 			return nil, err
 		}
 
-		portName := strings.Replace(tcpString, ":", "-", -1)
+		separatorPos := strings.Index(portsString, ":")
+		protocol := strings.ToUpper(portsString[0:separatorPos])
+		portName := strings.Replace(portsString[separatorPos+1:], ":", "-", -1)
```
Comment (xiangpengzhao) [inline pkg/kubectl/service_basic.go]: since we support specifying protocol, it'd be good to put protocol into the port name.

**Review by xiangpengzhao — CHANGES_REQUESTED** [review]

**Comment by sandoracs** [comment]
Thank you! I will work on it.

File: pkg/kubectl/service_basic.go (hunk 3 above)
Comment (sandoracs) [inline pkg/kubectl/service_basic.go]: so you suggest to remove protocol field from ServicePort struct?

Comment (xiangpengzhao) [inline pkg/kubectl/service_basic.go]: nope. I meant we might want to generate portName as `protocol-port-targetPort`, e.g. `tcp-456-9898`.

**Comment by sandoracs** [comment]
I've found an issue (https://github.com/kubernetes/kubectl/issues/35), which currently breaks testing.

**Comment by xiangpengzhao** [comment]
Hi @sandoracs , are you still working on this?

**Pull request closed without merging**